protocols/imap/src/main/java/org/apache/james/imap/decode/parser/GetAnnotationCommandParser.java

```diff
@@ -0,0 +1,147 @@
+                consumeKey(request, annotationRequest);
+            }
+
+            Preconditions.checkArgument(annotationRequest.isValidParams());
+            return annotationRequest;
+        } catch (NullPointerException e) {
+            throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, e.getMessage());
+        } catch (IllegalArgumentException e) {
+            throw new DecodingException(HumanReadableText.ILLEGAL_ARGUMENTS, e.getMessage());
+        }
+    }
+
+    private void consumeParenthesis(ImapRequestLineReader request, GetAnnotationRequest getAnnotationRequest) throws DecodingException {
+        request.consumeChar('(');
+        switch (request.nextChar()) {
+            case 'M':
+                getAnnotationRequest.setMaxsizeValue(consumeMaxsizeOpt(request));
+                break;
+
+            case 'D':
+                getAnnotationRequest.setDepthValue(consumeDepthOpt(request));
+                break;
+
+            default:
+                getAnnotationRequest.setKeys(consumeKeys(request));
+                break;
+        }
+        request.consumeChar(')');
+        if (!getAnnotationRequest.getKeys().isEmpty()) {
+            request.eol();
+        }
+    }
+
+    private Depth consumeDepthOpt(ImapRequestLineReader request) throws DecodingException {
+        request.consumeWord(new CharacterValidator() {
+            int pos = 0;
+            public boolean isValid(char chr) {
+                if (pos >= DEPTH.length) {
+                    return false;
+                } else {
+                    return DEPTH[pos++] == ImapRequestLineReader.cap(chr);
+                }
+            }
+        });
+        return Depth.fromString(request.atom());
+    }
+    public static void main(String[] agrs) {
+        System.out.println(Depth.fromString("-1"));
+    }
+
+    private long consumeMaxsizeOpt(ImapRequestLineReader request) throws DecodingException {
+        request.consumeWord(new CharacterValidator() {
+            int pos = 0;
+            public boolean isValid(char chr) {
+                if (pos >= MAXSIZE.length) {
+                    return false;
+                } else {
+                    return MAXSIZE[pos++] == ImapRequestLineReader.cap(chr);
+                }
+            }
+        });
+        return request.number(true);
+    }
+
+    private void consumeKey(ImapRequestLineReader request, GetAnnotationRequest getAnnotationRequest) throws DecodingException {
+        getAnnotationRequest.setKeys(ImmutableSet.of(request.atom()));
+        request.eol();
+    }
+
+    private Set<String> consumeKeys(ImapRequestLineReader request) throws DecodingException {
+        Builder<String> keys = ImmutableSet.builder();
+        do {
+            keys.add(request.atom());
+        } while (request.nextWordChar() != ')');
+
+        return keys.build();
+    }
+
+    private char nextNonSpaceChar(ImapRequestLineReader request) throws DecodingException {
```

Comment:
It seems usefull in other places. Can't you add it to ImapRequestLineReader instead ? it seems to be in the scope of this object responsabilities...